File: cpp/0560.0_Subarray_Sum_Equals_K.cpp
/*
Brute Force
Time: O(N^2)
Space: O(1)

85 / 89 个通过测试用例
状态：超出时间限制
提交时间：几秒前
*/

class Solution {
public:
    int subarraySum(vector<int>& nums, int k) {
        int count = 0;
        for (int end = 0; end < nums.size(); ++end) {
            int total = 0;
            for (int start = end; start >= 0; --start) {
                total += nums[start];
                if (total == k) {
                    count += 1;
                }
            }
        }
        return count;
    }
};



/*
approach: Prefix Sum + Hash Table (Two Sum)
Time: O(N)
Space: O(N)

执行用时：68 ms, 在所有 C++ 提交中击败了99.54%的用户
内存消耗：35.1 MB, 在所有 C++ 提交中击败了93.40%的用户
*/

class Solution {
public:
    int subarraySum(vector<int>& nums, int k) {
        int count = 0, pre = 0;
        unordered_map<int, int> mp;
        mp[0] = 1;

        for (auto x: nums) {
            pre += x;
            if (mp.find(pre - k) != mp.end()) {
                count += mp[pre - k];
            }
            mp[pre]++;
        }
        return count;
    }
};


/*
prefix sum + hash table
T: O(N)
S: O(N)
执行用时：60 ms, 在所有 C++ 提交中击败了86.24% 的用户
内存消耗：36.3 MB, 在所有 C++ 提交中击败了39.61% 的用户
通过测试用例：90 / 90
*/
class Solution {
public:
    int subarraySum(vector<int>& nums, int k) {
        unordered_map<int, int> precnt = {make_pair(0, 1)};
        int ans = 0, presum = 0;
        for (int x: nums) {
            presum += x;
            if (precnt.find(presum - k) != precnt.end()) {
                ans += precnt[presum - k];
            }
            precnt[presum] += 1;
        }
        return ans;
    }
};

File: java/0560.0_Subarray_Sum_Equals_K.java
/*
approach: Brute Force
Time: O(N^2)
Space: O(1)

执行用时：1117 ms, 在所有 Java 提交中击败了37.43%的用户
内存消耗：41.5 MB, 在所有 Java 提交中击败了27.42%的用户
*/
class Solution {
    public int subarraySum(int[] nums, int k) {
        int count = 0;

        for (int end = 0; end < nums.length; end++) {
            int total = 0;
            for (int start = end; start >= 0; start--) {
                total += nums[start];
                if (total == k) {
                    count += 1;
                }
            }
        }

        return count;
    }
}



/*
approach: Prefix Sum + Hash Table (Two Sum)
Time: O(N)
Space: O(N)

执行用时：30 ms, 在所有 Java 提交中击败了50.88%的用户
内存消耗：40.7 MB, 在所有 Java 提交中击败了86.70%的用户
*/
class Solution {
    public int subarraySum(int[] nums, int k) {
        int count = 0, pre = 0;
        HashMap <Integer, Integer> mp = new HashMap <> ();
        mp.put(0, 1);

        for (int i = 0; i < nums.length; i++) {
            pre += nums[i];
            if (mp.containsKey(pre - k)) {
                count += mp.get(pre - k);
            }
            mp.put(pre, mp.getOrDefault(pre, 0) + 1);
        }

        return count;
    }
}


/*
执行用时：25 ms, 在所有 Java 提交中击败了93.98%的用户
内存消耗：40.8 MB, 在所有 Java 提交中击败了78.03%的用户
*/

import java.util.HashMap;

class Solution {
    public int subarraySum(int[] nums, int k) {
        int count = 0, pre = 0;
        // key: 前缀和
        // value: key 对应的前缀和的个数, the frequency of prefix sum
        HashMap <Integer, Integer> preSumFreq = new HashMap<> ();
        preSumFreq.put(0, 1);
        
        for (int num: nums) {
            pre += num;
            if (preSumFreq.containsKey(pre - k)) {
                count += preSumFreq.get(pre - k);
            }
            preSumFreq.put(pre, preSumFreq.getOrDefault(pre, 0) + 1);
        }

        return count;
    }
}




